feat: add p/demo/authctx

moul opened the pull request:
See https://twitter.com/jaekwon/status/1713308824269144423

jaekwon commented:
interesting. so is the idea to get an "Auth" per person/account/realm/anything, and pass this around where the convention is the correct auth is expected to be passed in as context?

moul commented:
Yes.
The purpose is to serve as a preliminary authentication check for keys, realms, or delegations.
It also serves as a representation of an owner, storing additional metadata along with std.Address.

jaekwon commented:
1. This is interesting, nice start.
2. I'm not sure if this is what is needed for Books yet, but *somebody* is going to want something like this.
3. This falls under the category of "auth context passing" authentication pattern.
4. Maybe Auth should be called AuthContext. "Auth" by itself is too general? 
5. Should it be an interface? 

Regarding 5.... there is more security provided by a concrete impl rather than an interface, because the concrete impl by nature allows for more opinionated decisions. On the other hand... "auth context passing" patterns already assume that the caller knows what the caller is doing... and yet YET on the other hand, still, having concrete structures makes it less likely to program a bug.

I think we should start with a lowerCase struct impl and see what emerges, before jumping into "auth is an interface".

jaekwon commented:
related: https://github.com/gnolang/gno/issues/1245

jaekwon commented:
Also, isn't this better than permissions?

```go
if auth.IsAuthenticated(action) { 
   // perform the action
}
```

```go
// someperm comes from out of nowhere
if auth.HasPerm(someperm) {
    // perform some action independent of someperm
}
```

You have some separation of concerns between the action and this new type called a "permission". vs,

```go
func (_) IsAuthorized(action Action) error {
    switch action.(type) {
    case: WriteAction:
    case: ReadAction:
    default: return UnrecognizedError()
    }
}
```

And IsAuthorized could use "permissions" internally, or not, depending on what is natural.

jaekwon commented:
I guess either way, something is added, whether a Permission or an Action, 
but arguably it is GOOD(tm) to always serialize device/user actions into Action data,
but it isn't always required to use a Permission-type to authorize actions.

moul commented:
Updates:
-  Renamed `auth` to `authctx` and `Auth` to `AuthContext`.
-  Removed "*Perm" and introduced an improved `IsAuthorized` callback, which is better but still not flawless.
-  Introduced the `Scope` wrapper to facilitate the creation of a privileged-dropped `AuthContext` from an existing one.

jaekwon commented on examples/gno.land/p/demo/authctx/authctx.gno:
```diff
@@ -0,0 +1,162 @@
+// AuthContext is an interface that represents an authentication context.
+type AuthContext interface {
+	Addr() std.Address
+	String() string
```
Can we document what you wrote earlier about the requirements for these?

moul replied:
```suggestion
	// String returns a unique and human-readable identifier, typically in the format "<type>:<addr>[/opts]".
	String() string
```

moul replied:
`AuthContext` represents an owner, either direct (Orig/Prev) or indirect (Delegated). It's not a comprehensive ACL tool—this might be better suited for a separate library.

To initiate an `AuthContext`, you must be the original owner, achievable through the `Orig/Prev` constructors or by `Approving` another realm/account for future delegated authentication. We might introduce `authctx.IsNative(ctx)` to determine the origin of the `AuthContext` implementation.

`IsAuthorized` defines scope, not full access, mirroring JWTs. We opt for Golang's private fields for authenticity verification. Maybe HasScope() would be a better name than IsAuthorized.

A feature in `AuthContext` allows privilege adjustments. This explains why `IsAuthorized` defaults to true for direct contexts but needs a specified scope for upcoming delegations. Then you can re-Scope (drop privileges) on any AuthContext.

In essence, `AuthContext` isn't purely for ACLs. Manage ACLs with tools like `p/acl`. For delegated contexts, scope is as important as ACLs. While ACLs verify an owner's actions, `AuthContext.IsAuthorized` assesses the caller's rights to claim ownership in the present scenario.

Regarding potential paths forward (that I would like to experiment in the upcoming days): 
1. Removing `IsAuthorized`, making `AuthContext` a container, and placing checks during object creation.
2. Rename `IsAuthorized` with `HasScope`.
3. Enabling custom `IsAuthorized` checks through closures, allowing diverse filtering methods.
4. Refining the current `p/acl` library, making it interface-based, and employing that interface for `IsAuthorized`.

jaekwon commented on examples/gno.land/p/demo/authctx/authctx.gno:
```diff
@@ -0,0 +1,162 @@
+func NewOrigAuthContext() AuthContext {
+	return origAuthContext{addr: std.GetOrigCaller()}
+}
+func (a origAuthContext) Addr() std.Address               { return a.addr }
```
I would even skip implementing these until annealed but up to you!

jaekwon commented:
For next time, my two cents
I think it's easier to just discuss snippets of code rather than the whole implementation.
Struct and interface type defs and documentation, as if writing a basic tutorial.

jaekwon commented on examples/gno.land/p/demo/authctx/authctx.gno:
```diff
@@ -0,0 +1,163 @@
+}
+func (a origAuthContext) Addr() std.Address               { return a.addr }
+func (a origAuthContext) String() string                  { return "orig:" + string(a.Addr()) }
+func (a origAuthContext) IsAuthorized(action string) bool { return true }
```
How can an AuthContext authorize an action without knowing from whom the action comes from?
Is this supposed to return true all the time or is it a stub?

jaekwon replied:
"correct auth is expected to be passed in as context"
So I guess that makes sense, but something seems off.
An authorization should be of both the object/caller and the subject.
You can assume the caller (more on that below) but still there is no subject.
Upon what is the action authorized? 
It could be a naming issue, s/action/request/g for example would work, with request.action and request.subject.

Something about this general approach feels weird though.
Here's an analogy:
In a secure facility such as a building with access controls,
you have an identity card, which you swipe onto sensors, to open doors.

* You don't put the ACL data *in* the card, because this increases surface area a lot.
* There would be all kinds of ways to attack the card itself.
* Instead, usually the card merely holds an identity, and the ACL data is elsewhere in a central place.
* But AuthCtx is like an access control card that has the logic of permissioning inside locally.
* And the interfaces say that the way to use it is to implement your own access control logic card.

Wouldn't it make more sense to just pass around a Context which includes self identity, and to have a black box AuthControl.Authenticate(caller/subject, object, more action data)? 

* The user has less freedom, which is good -- less ways to mess up.
* The security is contained within the implementation of AuthControl, easier to audit.

I don't see the benefit of the assume-the-caller AuthCtx that is exposed to the user.

moul replied:
First, let me explain my initial idea for this package:

Think of `AuthContext` like `JWT` but simpler. It's a way to prove that something came from a trusted source and can be used for authentication. It allows you to pass around authentication information securely and preserve authenticity.

In a `JWT`, you have scopes or roles, but it's mainly used to manage permissions, not all the access control rules of an app.

Another analogy I like is sudo. When you're a regular user, you have limited permissions. But when you use sudo, you temporarily get more power. However, you can configure sudo to only give you specific powers for certain tasks (scopes). The final app still checks if you have the right permissions, and sudo is responsible to create the `AuthContext`.

`AuthContext` also has the advantage of being storable. Instead of just storing an address to represent an owner, you can store more context. This is useful when you want your system (like a book/boards) to be usable by different accounts, contracts, and delegates. No need for boilerplate code on your realm.

`Orig`/`Prev` AuthContexts have a default scope of "all," but when you delegate rights, you can limit them. There's also a helper to change (drop privileges) on any `AuthContext`, including `Orig`/`Prev` ones.

In short, `AuthContext` isn't just for access control lists (ACLs). Use tools like `p/acl` for that. For delegated contexts, scope is as important as ACLs. While ACLs verify an owner's actions, `AuthContext.IsAuthorized` checks the caller's rights in the current situation.

Now, I see two main directions:
1. Keep `AuthContext` as it is, making it better at context management, but not a complete ACL solution. It's more of a flexible authentication context for writing apps with less boilerplate.
2. Make AuthContext a complete ACL solution, removing the role system, and using an interface that can be implemented by different ACL mechanisms.

Possible next steps (things I plan to consider/hack on):
1. Introduce `authctx.IsNative(ctx)` to determine where the AuthContext comes from.
2. Remove ACLs, making AuthContext just a container, and add checks during object creation.
3. Expand AuthContext with features like expiration.
4. Allow custom `IsAuthorized` checks through closures.
5. Improve the current `p/acl` library and make it interface-based.
6. Rename `IsAuthorized` to `HasScope`.
7. Move the ACL system in authctx to an interface implementation, so it's still there but as a default option.

jaekwon replied:
I'll try to condense my general understanding here.

If it's a restrictor, then why not just fail early instead of wrapping contexts? For example, if Bob orig caller calls some realm logic (scoped to realm permissions) which calls performs some action on a Book, why not just fail early when the action can't be performed by Bob? And if Bob is allowed, then why not fail early when the realm logic is not allowed?

It seems to me this is merely delaying logic that could be performed immediately into a chain of callbacks, and what's the point of that? I don't see the benefit.

thehowl commented:
Closing as stale.
